shepherd/Codegen.py

```diff
@@ -0,0 +1,771 @@
+
+@memoize
+def limit_input_to(limit):
+    '''Generate a function to limit size of inputs'''
+    def retval(input_val):
+        while input_val > limit:
+            input_val = (input_val % limit) + (input_val // limit)
+        return input_val
+    return retval
+
+@memoize
+def compose_funcs(func_a, func_b):
+    '''
+    Composes two single-input functions together, A(B(x))
+    '''
+    def composed(input_x):
+        return func_a(func_b(input_x))
+    return composed
+
+
+def identity(x):
+    '''
+    Used only in the (hopefully) rare event that none of the other
+    functions are bijections with a given domain of RFIDs
+    '''
+    return x
+
+def get_function_mapping():
+    '''
+    Returns a dictionary that maps code digits to the function to run
+    on the RFID values. Always returns values which are equal to each other
+    by value. It is safe to mutate the return value.
+    >>> get_function_mapping() == get_function_mapping()
+    True
+    '''
+
+    func_map = {}
+    func_map[0] = identity
+    func_map[1] = next_power
+    func_map[2] = reverse_digits
+    func_map[3] = compose_funcs(smallest_prime_fact, limit_input_to(1000000))
+    func_map[4] = double_caesar_cipher
+    func_map[5] = silly_base_two
+    func_map[6] = most_common_digit
+    func_map[7] = valid_isbn_ten
+    func_map[8] = simd_four_square
+
+    return func_map
+
+def staff_decode_precompute(rfids):
+    '''
+    To speed up calls to staff_decode over the same domain of known RFIDs, we
+    can run this precomputation and pass the return value to the `staff_decode`
+    function as the `precompute` argument.
+
+    Returns a dictionary of dictionaries, where the first-level key is the digit
+    corresponding with that student function, and the second-level key is the
+    rfid we wish to evalute the student function on.
+    '''
+    applied_map = {} # Precomputation to use for speeding up staff_decode calls
+    func_map = get_function_mapping()
+
+    # Apply every function to every rfid. This precomputation step
+    # allows us to memoize the potentially expensive example solutions for the
+    # student functions.
+    for i in func_map:
+        applied_map[i] = {}
+        for rfid in rfids:
+            applied_map[i][rfid] = func_map[i](rfid)
+    return applied_map
+
+def staff_decode(challenge_code, rfid_seed, precompute=None):
+    '''
+    Staff solution for the student decoder. Takes as input a code and the data
+    collected by an RFID scanner, and outputs the result of applying the
+    encoded functions onto that RFID "seed".
+    '''
+    if precompute is None:
+        func_map = get_function_mapping()
+
+    code = challenge_code
+    output = ''
+    while code > 0:
+        digit = code % 10
+        code //= 10
+
+        if precompute is None:
+            result = func_map[digit](rfid_seed)
+        else:
+            result = precompute[digit][rfid_seed]
+
+        output += str(result)
+    output = int(output) % 10000
```

Comment:
Actually I think we don't want to mod this number by 10000. This would mean only the least significant digits would be used, so the first few function calls could be useless. Am pretty sure actually the random.seed input can be like any number.